### PR TITLE
[Access] add signer ids to access API

### PR DESCRIPTION
**Pull request by zhangchiqing**
This PR populates the [parent_voter_ids](https://github.com/onflow/flow/blob/ad0378b94e44b6ddfac57cc5bded14f9865c3c0a/protobuf/flow/entities/block_header.proto#L17) in the Access API’s BlockHeader

### CODE REVIEW

**Review by jordanschalm — APPROVED** [review]

**Review by AlexHentschel — APPROVED** [review]
Looks good. Have a few structural recommendations. My only concern is the re-implementation of the logic to select the authorized consensus members (-> [comment](https://github.com/onflow/flow-go/pull/2762#discussion_r917085129))

Side-comment, which we could fix in this PR: 
* I feel that the function to convert a byte slice to an Identifier should check that the length matches exactly the expected `IdentifierLen` (32 bytes): https://github.com/onflow/flow-go/blob/19f6a66905d97c29dfbb71e11ad1e63331f54886/engine/common/rpc/convert/validate.go#L29-L34

File: state/protocol/util.go [hunk 1]
```diff
@@ -96,3 +96,26 @@ func FindGuarantors(state State, guarantee *flow.CollectionGuarantee) ([]flow.Id
 
 	return guarantorIDs, nil
 }
+
+// DecodeSignerIDs decodes the signer indices from the given block header, and finds the signer identifiers from protocol state
+// Expected Error returns during normal operations:
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/util.go:101]: 🤔 I feel that this is an unnecessary restriction. Would be better, if this function could decode signer indices for _any_ block (main chain or collector cluster). I think we could make that possible, i.e. the function generally applicable, by using `hotstuff.Committee` as in input and not `protocol.State`. I feel this is the specific reason why we have the `hotstuff.Committee` as an abstraction.

[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/util.go:101]: That's fair, the suggestion is based on the implementation, and the fact that this is in the `state/protocol` package. If we use `Committee` as input, the function can be generic to either main chain or cluster chain 👍

File: state/protocol/util.go [hunk 2]
```diff
@@ -96,3 +96,26 @@ func FindGuarantors(state State, guarantee *flow.CollectionGuarantee) ([]flow.Id
...
+//  * storage.ErrNotFound if block not found for the given header
+//  * signature.InvalidSignerIndicesError if `signerIndices` does not encode a valid set of consensus committiee
+func DecodeSignerIDs(state State, header *flow.Header) ([]flow.Identifier, error) {
+	// root block does not have signer indices
+	if header.ParentVoterIndices == nil && header.View == 0 {
+		return []flow.Identifier{}, nil
+	}
```
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply state/protocol/util.go:108]: Then we assume the block has passed the validation, right?

File: access/handler.go [hunk 3]
```diff
@@ -56,7 +59,12 @@ func (h *Handler) GetLatestBlockHeader(
 		return nil, err
 	}
 
-	return blockHeaderResponse(header)
+	signerIDs, err := protocol.DecodeSignerIDs(h.state, header)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply access/handler.go:65]: blockHeaderResponse is a pure function. We would need to pass the protocol state instance into that function as well.

File: state/protocol/util.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/util.go:109]: Totally agree that the AN should only answer queries for blocks that it knows are valid. Nevertheless, I think this is not the case at the moment: the AN just looks up the block in its data base and answers the query if it finds the respective block (header) -- irrespective of the block being valid or not. However, adding the check for block validity is probably beyond the scope of this PR. 

Overall, our root blocks currently _always_ have view 0, despite this not being required by the protocol. You are right that it is better to return an error as opposed to answering a query with an invalid block.

**Comment by AlexHentschel** [comment]
added revisions from draft PR https://github.com/onflow/flow-go/pull/2789 here

**Comment by AlexHentschel** [comment]
bors merge

**Pull request merged**